### PR TITLE
Adyen: Allow for executeThreeD to be passed as false

### PR DESCRIPTION
**Pull request by naashton**
Passing false to `executeThreeD` is an option for this gateway that some
users may want

CE-550

Unit: 66 tests, 315 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

Remote: 89 tests, 343 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

### CODE REVIEW

File: lib/active_merchant/billing/gateways/adyen.rb [hunk 1]
```diff
@@ -418,10 +418,10 @@ def add_3ds(post, options)
             post[:additionalData][:scaExemption] = options[:sca_exemption] if options[:sca_exemption]
           end
         else
-          return unless options[:execute_threed] || options[:threed_dynamic]
+          return unless !options[:execute_threed].nil? || !options[:threed_dynamic].nil?
```
Comment (molbrown) [inline lib/active_merchant/billing/gateways/adyen.rb:421]: `browserInfo` should only be included if execute_threed or threed_dynamic is true:
- `browserInfo` is required if `executeThreeD` is passed as true
- the inclusion of `browserInfo` is what tells Adyen to check Dynamic 3DS rules and maybe perform 3DS https://docs.adyen.com/risk-management/dynamic-3d-secure#page-introduction

Comment (naashton) [reply lib/active_merchant/billing/gateways/adyen.rb:421]: Alright, I think I've got what we want here. I was mostly following this https://docs.adyen.com/payments-fundamentals/psd2-sca-compliance-and-implementation-guide#specifypreferenceinyourapirequest for `executeThreeD` situations.

**Comment by naashton** [comment]
I didn't add a unit test because it looked like all of the bases were covered there.

File: test/remote/gateways/remote_adyen_test.rb [hunk 2]
```diff
@@ -180,6 +180,13 @@ def test_successful_authorize_with_3ds
     refute response.params['paRequest'].blank?
   end
 
+  def test_successful_authorize_with_execute_threed_false
+    assert response = @gateway.authorize(@amount, @three_ds_enrolled_card, @options.merge(execute_threed: false, sca_exemption: 'lowValue'))
+    assert response.test?
+    refute response.authorization.blank?
+    assert_equal response.params['resultCode'], 'RedirectShopper'
```
[the remark this answers is not in the document]
Comment (naashton) [reply test/remote/gateways/remote_adyen_test.rb:187]: I'm not entirely sure what the response is supposed to be, but this test ends up returning this, `@message="Received unexpected 3DS authentication response. Use the execute_threed and/or threed_dynamic options to initiate a proper 3DS flow.`. Unfortunately I can't find anything in the documentation related to this message.

[the remark this answers is not in the document]
Comment (molbrown) [reply test/remote/gateways/remote_adyen_test.rb:187]: Running that test with `DEBUG_ACTIVE_MERCHANT=true` shows that executeThreed is not added to the request:

{"merchantAccount":"SpreedlyCOM294","reference":"123","amount":{"value":"100","currency":"USD"},"card":{"expiryMonth":10,"expiryYear":2020,"holderName":"Longbob Longsen","number":"4917610000000000","cvc":"737"},"telephoneNumber":"(555)555-5555","shopperEmail":"john.smith@test.com","shopperIP":"77.110.174.153","additionalData":{},"shopperReference":"John Smith","shopperInteraction":"Ecommerce","recurringProcessingModel":"CardOnFile","billingAddress":{"street":"456 My Street","houseNumberOrName":"Apt 1","postalCode":"K1C2N6","city":"Ottawa","stateOrProvince":"ON","country":"CA"}}

File: lib/active_merchant/billing/gateways/adyen.rb [hunk 3]
```diff
@@ -418,10 +418,10 @@ def add_3ds(post, options)
...
 
-          post[:browserInfo] = { userAgent: options[:user_agent], acceptHeader: options[:accept_header] }
-          post[:additionalData] = { executeThreeD: 'true' } if options[:execute_threed]
+          post[:browserInfo] = { userAgent: options[:user_agent], acceptHeader: options[:accept_header] } if options[:execute_threed] || options[:threed_dynamic]
+          post[:additionalData] = { executeThreeD: options[:execute_threed] } if options[:execute_threed]
```
Comment (molbrown) [inline lib/active_merchant/billing/gateways/adyen.rb:424]: `if options[:execute_threed]` evaluates to false here, so the field is not added. I tested using the !nil? pattern you did on line 421, which works.

Comment (naashton) [reply lib/active_merchant/billing/gateways/adyen.rb:424]: Oh right, I keep assuming `if options[:execute_threed]` is going to return truthy if the value is present.

**Comment by naashton** [comment]
Woo! Thanks for walking me through this!

**Pull request closed without merging**